File: UClient/Assets/UFramework/Scripts/Core/Logger/Logger.cs
/*
 * @Author: fasthro
 * @Date: 2020-11-25 12:26:03
 * @Description: logger
 */

using System.Text;

namespace UFramework
{
    public enum LogLevel
    {
        Debug = 0,
        Info,
        Warn,
        Error,
        Fatal,
    }

    public class Logger
    {
        readonly static int LEVEL_DEBUG = (int)LogLevel.Debug;
        readonly static int LEVEL_INFO = (int)LogLevel.Info;
        readonly static int LEVEL_WARN = (int)LogLevel.Warn;
        readonly static int LEVEL_ERROR = (int)LogLevel.Error;
        readonly static int LEVEL_FATAL = (int)LogLevel.Fatal;

        readonly static string[] LOG_LEVEL_TITLES = new string[] { "DEBUG", "INFO", "WARN", "ERROR", "FATAL" };

        static int CUR_LEVEL = LEVEL_DEBUG;

        public static void Debug(object message)
        {
            if (CheckLevel(LEVEL_DEBUG))
                UnityEngine.Debug.Log(string.Format("<color=#04fd14>[{0}]</color> {1}", LOG_LEVEL_TITLES[LEVEL_DEBUG], message.ToString()));
        }

        public static void Info(object message)
        {
            if (CheckLevel(LEVEL_INFO))
                UnityEngine.Debug.Log(string.Format("<color=#49c9c1>[{0}]</color> {1}", LOG_LEVEL_TITLES[LEVEL_INFO], message.ToString()));
        }

        public static void Error(object message)
        {
            if (CheckLevel(LEVEL_ERROR))
                UnityEngine.Debug.Log(string.Format("<color=#ff0417>[{0}]</color> {1}", LOG_LEVEL_TITLES[LEVEL_ERROR], message.ToString()));
        }

        public static void Wraning(object message)
        {
            if (CheckLevel(LEVEL_WARN))
                UnityEngine.Debug.Log(string.Format("<color=#ebff28>[{0}]</color> {1}", LOG_LEVEL_TITLES[LEVEL_WARN], message.ToString()));
        }

        public static void SetLevel(LogLevel logLevel)
        {
            CUR_LEVEL = (int)logLevel;
        }

        public static int GetLevel()
        {
            return CUR_LEVEL;
        }

        static bool CheckLevel(int level)
        {
            return level >= CUR_LEVEL;
        }

        public static void DebugBytes(byte[] bytes)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("[");
            for (int i = 0; i < bytes.Length; i++)
            {
                sb.Append(bytes[i]);
                if (i < bytes.Length - 1)
                {
                    sb.Append(",");
                }
            }
            sb.Append("] size " + bytes.Length);

            Debug(sb.ToString());
        }
    }
}

File: UClient/Assets/UFramework/Scripts/Lua/Logger/Logger.lua
--[[
Author: fasthro
Date: 2020-08-28 12:15:46
Description: log
--]]
local logging = require "Logger.Logging"

local prepare = function(pattern, level, message)
    message = string.gsub(message, "%%", "%%%%")
    pattern = string.gsub(pattern, "%%level", level)
    pattern = string.gsub(pattern, "%%message", message)
    return pattern
end

local _logdebug =
    logging.new(
    function(self, level, message)
        local pattern = "<color=#04fd14>[%level]</color> %message"
        UnityEngine.Debug.Log(prepare(pattern, level, message))
        return true
    end
)

local _loginfo =
    logging.new(
    function(self, level, message)
        local pattern = "<color=#49c9c1>[%level]</color> %message"
        UnityEngine.Debug.Log(prepare(pattern, level, message))
        return true
    end
)

local _logerror =
    logging.new(
    function(self, level, message)
        local pattern = "<color=#ff0417>[%level]</color> %message"
        UnityEngine.Debug.LogError(prepare(pattern, level, message))
        return true
    end
)

local _logwarning =
    logging.new(
    function(self, level, message)
        local pattern = "<color=#ebff28>[%level]</color> %message"
        UnityEngine.Debug.LogWarning(prepare(pattern, level, message))
        return true
    end
)

-- 日志等级(与C#日志等级同步)
local LEVEL_DEBUG = 0
local LEVEL_INFO = 1
local LEVEL_WARN = 2
local LEVEL_ERROR = 3
local LEVEL_FATAL = 4

-- 当前日志等级
local CUR_LOG_LEVEL = 0

-- 日志等级检查
local checklevel = function(level)
    return level >= CUR_LOG_LEVEL
end

local logger = {}

-- log
function logger.debug(...)
    if checklevel(LEVEL_DEBUG) then
        _logdebug:debug(...)
    end
end

-- info
function logger.info(...)
    if checklevel(LEVEL_INFO) then
        _loginfo:info(...)
    end
end

-- error
function logger.error(...)
    if checklevel(LEVEL_ERROR) then
        _logerror:error(...)
    end
end

-- warning
function logger.warning(...)
    if checklevel(LEVEL_WARN) then
        _logwarning:warn(...)
    end
end

-- set level
function logger.setlevel(level)
    CUR_LOG_LEVEL = level
end

-- tostring
function logger.tostring(value)
    return logging.tostring(value)
end

-- buffer tostring
function logger.buffer_tostring(buffer)
    local res = "["
    local len = #buffer
    for k = 1, len do
        res = res .. buffer:byte(k)
        if k ~= len then
            res = res .. ","
        end
    end
    res = res .. "] size: " .. len
    return res
end

_G.logger = logger
